File: app/src/main/java/me/devsaki/hentoid/json/sources/PixivSeriesIllustMetadata.java
package me.devsaki.hentoid.json.sources;

import static me.devsaki.hentoid.parsers.ParseHelperKt.cleanup;

import com.annimon.stream.Stream;

import java.util.ArrayList;
import java.util.List;

import kotlin.Pair;
import me.devsaki.hentoid.database.domains.Chapter;
import me.devsaki.hentoid.enums.Site;

/**
 * Data structure for Pixiv's "series-content" mobile endpoint
 */
@SuppressWarnings({"unused, MismatchedQueryAndUpdateOfCollection", "squid:S1172", "squid:S1068"})
public class PixivSeriesIllustMetadata {

    private Boolean error;
    private String message;
    private SeriesContentBody body;

    private static class SeriesContentBody {
        private List<SeriesContent> series_contents;
    }

    private static class SeriesContent {
        private String id;
        private String title;

        public String getId() {
            return id;
        }

        public String getTitle() {
            return title;
        }
    }

    public boolean isError() {
        return error;
    }

    public String getMessage() {
        return message;
    }

    public List<Pair<String, String>> getIllustIdTitles() {
        List<Pair<String, String>> result = new ArrayList<>();
        if (null == body || null == body.series_contents) return result;
        return Stream.of(body.series_contents).map(c -> new Pair<>(c.getId(), c.getTitle())).toList();
    }

    public List<Chapter> getChapters(long contentId) {
        List<Chapter> result = new ArrayList<>();
        if (null == body || null == body.series_contents) return result;
        int order = 0;
        for (SeriesContent sc : body.series_contents) {
            String forgedUrl = Site.PIXIV.getUrl() + "artworks/" + sc.getId();
            Chapter chp = new Chapter(order++, forgedUrl, cleanup(sc.getTitle()));
            chp.setUniqueId(sc.getId());
            chp.setContentId(contentId);
            result.add(chp);
        }
        return result;
    }
}

File: app/src/main/java/me/devsaki/hentoid/json/sources/pixiv/PixivSeriesIllustMetadata.kt
package me.devsaki.hentoid.json.sources.pixiv

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import me.devsaki.hentoid.database.domains.Chapter
import me.devsaki.hentoid.enums.Site
import me.devsaki.hentoid.parsers.cleanup

/**
 * Data structure for Pixiv's "series-content" mobile endpoint
 */
@JsonClass(generateAdapter = true)
data class PixivSeriesIllustMetadata(
    val error: Boolean? = null,
    val message: String? = null,
    val body: SeriesContentBody? = null
) {
    fun getIllustIdTitles(): List<Pair<String, String>> {
        if (body?.seriesContents == null) return emptyList()
        return body.seriesContents.map { Pair(it.id ?: "", it.title ?: "") }
    }

    fun getChapters(contentId: Long): List<Chapter> {
        val result: MutableList<Chapter> = ArrayList()
        if (body?.seriesContents == null) return result
        for ((order, sc) in body.seriesContents.withIndex()) {
            val forgedUrl = Site.PIXIV.url + "artworks/" + (sc.id ?: "")
            val chp = Chapter(order, forgedUrl, cleanup(sc.title))
            chp.uniqueId = sc.id ?: ""
            chp.setContentId(contentId)
            result.add(chp)
        }
        return result
    }
}

@JsonClass(generateAdapter = true)
data class SeriesContentBody(
    @Json(name = "series_contents")
    val seriesContents: List<SeriesContent>? = null
)

@JsonClass(generateAdapter = true)
data class SeriesContent(
    val id: String? = null,
    val title: String? = null
)